46_revocation_severity_and_context.md

```diff
@@ -0,0 +1,301 @@
+-->
+
+## Summary
+This enhancement proposal adds the capability differentiate and classify
+revocation events by adding severity levels and user defined or generated
+context.
+
+## Motivation
+Currently Keylime operates in a binary state. Either a device verified or not.
+
+Not all revocation events have to be handled with an equal severity. 
+Depending on the use case a failing PCR caused by a firmware update is handled 
+differently than IMA policy failures.
+
+Granular reporting of revocation events allows the user to make more informed 
+decisions if failure occurs. The additional provided context also helps with 
+reconstructing why a revocation was issued.
+
+This enhancement allows users to use Keylime in new scenarios like exam
+environments where a person must make an informed decision if a device is
+actually compromised.
+
+### Goals
+ * Allow the user to specify a severity level and context information for every
+   revocation event that is issued by Keylime.
+  * Tag all parts that cause revocations in Keylime with an unique id.
+  * Assign context provided by the user to (static) revocation events.
+  * Provide the option for Keylime to add context itself to revocation events.
+    This is useful if for example Eventlog parsing gains support for dynamic
+    policies.
+  * Extend the verifier API and tenant to add those rules to Keylime.
+ * Provide API entry point for checking which revocations events were triggered
+   for a agent.
+
+### Non-Goals
+ * The classification of revocation events is highly environment and
+   configuration specific. It is not a goal to provide a default configuration
+   other than all events are classified with the highest severity.
+
+
+## Proposal
+Currently revocation events are either send if something with the communication
+the the agent went wrong, the quote is not valid or one of the PCR based checks
+fail.
+
+This proposal changes that behavior and introduces the concept of severity
+levels of events and adds the option to provide more context for events.
+
+All parts that cause revocation events are tagged with an unique id and against
+those ids user scan specify a severity level and provide some context information.
+
+Instead of stop polling an agent if a failure occurred the agent is added back
+for checking if the failure is recoverable. If it is not a revocation event with
+the highest severity is generated.
+
+All the send events are visible in the verifier API for other systems to use.
+
+Keylime also gains the ability to specify context for a revocation event.
+
+### User Stories (optional)
+
+<!--
+Detail the things that people will be able to do if this enhancement is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system.  The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
```

Comment:
We should have a couple of user stories here just to be complete and make sure everyone is on the same page with the capabilities and problems being solved.

Reply:
Are now added